config: set buildkit gc enabled to default to true

jsternberg opened the pull request:
This will use the default settings for buildkit gc unless explicitly
disabled by setting `enabled: false` in the gc configuration.

Fixes #49826.

```markdown changelog
containerd image store: Enable BuildKit garbage collector by default.
```

thaJeztah commented on daemon/config/builder_test.go:
```diff
@@ -89,3 +89,12 @@ func TestBuilderGCFilterUnmarshal(t *testing.T) {
 	assert.DeepEqual(t, cfg.Policy, expectedPolicy, cmp.AllowUnexported(BuilderGCFilter{}))
 }
+
+func TestBuilderGC_EnabledByDefault(t *testing.T) {
+	tempFile := fs.NewFile(t, "config", fs.WithContent(`{}`))
+	configFile := tempFile.Path()
+
+	cfg, err := MergeDaemonConfigurations(&Config{}, nil, configFile)
+	assert.NilError(t, err)
+	assert.Assert(t, cfg.Builder.GC.IsEnabled())
+}
```
I don't think have a test that tests explicit "disabled"; perhaps some basic table-test for this one with some permutations of builder, and gc config being (partially) set?

```go
func TestBuilderGC_Enabled(t *testing.T) {
	tests := []struct {
		doc, config string
		expected    bool
	}{
		{doc: "empty config", config: ``, expected: true},
		{doc: "empty json", config: `{}`, expected: true},
		{doc: "empty builder", config: `{"builder": {}}`, expected: true},
		{doc: "empty gc", config: `{"builder": {"gc": {}}}`, expected: true},
		{doc: "gc enabled", config: `{"builder": {"gc": {"enabled": true}}}`, expected: true},
		{doc: "gc disabled", config: `{"builder": {"gc": {"enabled": false}}}`, expected: false},
		{doc: "gc with policy", config: `{"builder": {"gc": {"policy": []}}}`, expected: true},
	}
	for _, tc := range tests {
		t.Run(tc.doc, func(t *testing.T) {
			tempFile := fs.NewFile(t, "config", fs.WithContent(tc.config))
			configFile := tempFile.Path()

			cfg, err := MergeDaemonConfigurations(&Config{}, nil, configFile)
			assert.NilError(t, err)
			assert.Equal(t, cfg.Builder.GC.IsEnabled(), tc.expected)
		})
	}
}
```

jsternberg replied:
I copied and pasted this code section directly since I couldn't think of a better way to write it.

thaJeztah commented on daemon/config/builder.go:
```diff
@@ -84,13 +84,17 @@ func (x *BuilderGCFilter) UnmarshalJSON(data []byte) error {
 	DefaultMaxUsedSpace  string          `json:",omitempty"`
 	DefaultMinFreeSpace  string          `json:",omitempty"`
 }
 
+func (x *BuilderGCConfig) IsEnabled() bool {
+	return x.Enabled == nil || *x.Enabled
+}
+
 func (x *BuilderGCConfig) UnmarshalJSON(data []byte) error {
 	var xx struct {
 		Enabled              bool            `json:",omitempty"`
```
I was even considering if we had to modify `BuilderGCConfig. Enabled` to be a pointer, or if we'd only need to modify this temporary struct (unmarshal to xx` and if it's nil, then we set the default to true, which we already do below), but that would require `New()` to set the default (for cases where the `UnmarshalJSON` is not called, and looks like we'd have to patch other places to start with `New()` (default config) instead of starting with an empty config.

jsternberg replied:
Modifying the temporary struct doesn't work because `UnmarshalJSON` only gets called _if_ there's a gc section. So doing the change in the temporary struct works for this:

```
"builder": {"gc": {}}
```

But it doesn't work when `"gc"` is missing or if the entire configuration is missing. At the same time, doing it this way ensures that `"enabled": null` isn't accidentally an acceptable value. If the temporary struct has a pointer, then it would be ok for someone to set `null` as the value. If we set `Enabled` to true before running unmarshal then it will be set to true if the attribute isn't present otherwise it'll be overridden and `null` will cause it to error.

thaJeztah replied:
Yeah, I came to that conclusion as well; I had a bit of a brain-fart and thought it would be called unconditionally, but that's that didn't really make sense;

> for cases where the UnmarshalJSON is not called

jsternberg committed:
config: set buildkit gc enabled to default to true

This will use the default settings for buildkit gc unless explicitly
disabled by setting `enabled: false` in the gc configuration.

Signed-off-by: Jonathan A. Sternberg <jonathan.sternberg@docker.com>

thaJeztah reviewed (approved):
LGTM
